src/networkaddress.rs

```diff
@@ -0,0 +1,318 @@
+    #[doc(hidden)]
+    fn leave_multicast(
+        &self,
+        socket: RawFd,
+        interface: InterfaceName,
+        _token: PrivateToken,
+    ) -> std::io::Result<()>;
+}
+
+impl SealedNA for SocketAddrV4 {}
+
+impl NetworkAddress for SocketAddrV4 {
+    fn to_sockaddr(&self, _token: PrivateToken) -> libc::sockaddr_storage {
+        const _: () = assert!(
+            std::mem::size_of::<libc::sockaddr_storage>()
+                >= std::mem::size_of::<libc::sockaddr_in>()
+        );
+        const _: () = assert!(
+            std::mem::align_of::<libc::sockaddr_storage>()
+                >= std::mem::align_of::<libc::sockaddr_in>()
+        );
+
+        let mut result = zeroed_sockaddr_storage();
+        // Safety: the above assertions guarantee that alignment and size are correct.
+        // the resulting reference won't outlast the function, and result lives the entire
+        // duration of the function
+        let out = unsafe { &mut (*(&mut result as *mut _ as *mut libc::sockaddr_in)) };
+        out.sin_family = libc::AF_INET as _;
+        out.sin_port = u16::from_ne_bytes(self.port().to_be_bytes());
+        out.sin_addr = libc::in_addr {
+            s_addr: u32::from_ne_bytes(self.ip().octets()),
+        };
+
+        result
+    }
+
+    fn from_sockaddr(addr: libc::sockaddr_storage, _token: PrivateToken) -> Option<Self> {
+        const _: () = assert!(
+            std::mem::size_of::<libc::sockaddr_storage>()
+                >= std::mem::size_of::<libc::sockaddr_in>()
+        );
+        const _: () = assert!(
+            std::mem::align_of::<libc::sockaddr_storage>()
+                >= std::mem::align_of::<libc::sockaddr_in>()
+        );
+
+        if addr.ss_family != libc::AF_INET as _ {
+            return None;
+        }
+
+        // Safety: the above assertions guarantee that alignment and size are correct
+        // the resulting reference won't outlast the function, and addr lives the entire
+        // duration of the function
+        let input = unsafe { &(*(&addr as *const _ as *const libc::sockaddr_in)) };
+        Some(SocketAddrV4::new(
+            Ipv4Addr::from(input.sin_addr.s_addr.to_ne_bytes()),
+            u16::from_be_bytes(input.sin_port.to_ne_bytes()),
+        ))
+    }
+}
+
+impl SealedMC for SocketAddrV4 {}
+
+impl MulticastJoinable for SocketAddrV4 {
+    fn join_multicast(
+        &self,
+        socket: RawFd,
+        interface: InterfaceName,
+        _token: PrivateToken,
+    ) -> std::io::Result<()> {
+        let request = libc::ip_mreqn {
+            imr_multiaddr: libc::in_addr {
+                s_addr: u32::from_ne_bytes(self.ip().octets()),
+            },
+            imr_address: libc::in_addr {
+                s_addr: u32::from_ne_bytes(Ipv4Addr::UNSPECIFIED.octets()),
+            },
+            imr_ifindex: InterfaceDescriptor {
+                interface_name: Some(interface),
+                mode: crate::interface::LinuxNetworkMode::Ipv6,
```

Comment:
I'm assuming this is an oversight? or is there a reason to use ipv6 here?

Reply:
woeps